### ISSUE TITLE
Error handling causes customer data leak

### ISSUE THREAD

**Issue by ryanpalmerweb** [magento/magento2#23122]
### Preconditions (*)
1. Magento 2.3.1

### Steps to reproduce (*)
1. Asynchronous email sending must be enabled in global settings
2. Backlog of unsent emails (e.g in Magento\Sales\Model\Order\Email\Sender\CreditmemoSender)
3. An error must be encountered during the getTransport method (e.g a plugin that references this tries to call an undefined function)
4. The application must run via cron so that any \Errors are automatically converted to \Exceptions by Magento\Framework\App\ErrorHandler.php

### Expected result (*)
The class Magento\Framework\Mail\Template\TransportBuilder is implemented using the Singleton pattern, so should reset its internal variables after an error to prevent customers from seeing other customers order details.

### Actual result (*)
The class Magento\Framework\Mail\Template\TransportBuilder is not reset, causing the customer who experienced the original issue to receive a copy of the next customers order/invoice/credit memo email.

### Background information ###

One of our clients was receiving complaints from customers that they were being emailed the wrong customers orders, and were therefore concerned about their own privacy.

We isolated the error and found that the Fooman EmailAttachments module was referencing the createAttachment function which existed in Magento 2.2 but was removed in Magento 2.3

In Magento 2.2 :
class Message extends \Zend_mail implements MessageInterface

In Magento 2.3 :
class Message implements MailMessageInterface

Searching the web for "Call to undefined method Magento\Framework\Mail\Message\Interceptor::createAttachment()" showed 68 results, meaning that this is not an isolated fault.

The probability of this happening to other installations as a result of breaking changes to the core is very high

### Suggested fix ###

The potential for a third party module to cause Magento to email the wrong customer is caused by the error handling in the SenderBuilder not resetting the TransportBuilder.

The send() function of the SenderBuilder should catch any errors, and reset the TransportBuilder to prevent this from happening.

Alternatively the TransportBuilder could be converted from a Singleton into a Factory.

**Comment by ryanpalmerweb** [comment]
Just to add to this, one of our clients was receiving complaints from customers that they were being emailed other customers orders, and were concerned about their own privacy.

We isolated the error and found that the Fooman EmailAttachments module was referencing the createAttachment function which existed in Magento 2.2 but was removed in Magento 2.3

In Magento 2.2 : 
class Message extends \Zend_mail implements MessageInterface

In Magento 2.3 :  
class Message implements MailMessageInterface

Therefore the probability of this happening to other installations as a result of breaking changes to the core is very high.

Googling for "Call to undefined method Magento\Framework\Mail\Message\Interceptor::createAttachment()" showed 68 results, meaning that this is not an isolated fault.

**Comment by ryanpalmerweb** [comment]
@okorshenko I can see this change relates to MAGETWO-67680

**Comment by orlangur** [comment]
```
Suggested fix
The potential for a third party module to cause Magento to email the wrong customer is caused by the error handling in the SenderBuilder not resetting the TransportBuilder.

The send() function of the SenderBuilder should catch any errors, and reset the TransportBuilder to prevent this from happening.

Alternatively the TransportBuilder could be converted from a Singleton into a Factory.
```

Core code must not take into consideration buggy plugins/interceptors. Sounds like the bug is in Fooman module.

**Comment by ryanpalmerweb** [comment]
It's surely dangerous to assert that there could never be an exception in this area of code, given the consequences.

In this case, a backwards incompatible change to the Magento core (a function being removed) led to the issue being triggered by a 3rd party plugin, however *any* exception in this area of code could lead to customers being inadvertently emailed and it's quite a serious breach of customer privacy if this happens.

> "A personal data breach may, if not addressed in an appropriate and timely manner, result in physical, material or non-material damage to natural persons such as loss of control over their personal data or limitation of their rights, discrimination, identity theft or fraud, financial loss, unauthorised reversal of pseudonymisation, damage to reputation, loss of confidentiality of personal data protected by professional secrecy or any other significant economic or social disadvantage to the natural person concerned."

Generally, pushing customer data into a singleton object and relying on a specific execution path to reset this for the next use (which is bypassed by any exception handling) isn't safe behaviour

It would be relatively easy to safeguard this area of code, by initially resetting the transportBuilders internals or retrieving it via a factory pattern - is there any reason that this can't / shouldn't be done?

**Comment by orlangur** [comment]
@ryanpalmerweb no objections to re-create `TransportBuilder` each time, just that there is no defect in core currently which needs to be fixed. Third-party extensions need to be tested properly when upgrading.

**Comment by fooman** [comment]
As I was mentioned in here I wanted to note for future people coming across this that the issue would not exhibit when running the latest released versions of our extensions (and Composer should have prevented the 2.3.1 upgrade without also upgrading our extension) but I fully support making this code more defensive.

**Comment by ihor-sviziev** [comment]
@engcom-Golf could you check if this issue still valid on 2.4-develop?

**Comment by ihor-sviziev** [comment]
@engcom-Bravo maybe you could check this issue?

**Comment by hostep** [comment]
Let's keep this issue open, it sounds important enough. Would be great if somebody could verify and assign priorities. @sidolov @gabrieldagama

**Comment by ihor-sviziev** [comment]
@gabrieldagama @sivaschenko, I think the priority for this issue should be set to p2, or even to p1.

**Comment by gabrieldagama** [comment]
Hi @ihor-sviziev @hostep, thanks for highlighting this issue. I wonder if this is still reproducible on 2.4-develop, as far as I can see the `getTransport()` method is always resetting its values since 2019.

```
    public function getTransport()
    {
        try {
            $this->prepareMessage();
            $mailTransport = $this->mailTransportFactory->create(['message' => clone $this->message]);
        } finally {
            $this->reset();
        }

        return $mailTransport;
    }
```

@ryanpalmerweb were you able to reproduce this issue on the latest release 2.4.2? Thanks!

**Comment by ryanpalmerweb** [comment]
Hi @gabrieldagama,

Looking at the commit below the issue looks to be fixed in 2.3.3-p1 and 2.4.2, although I'm not able to check at the moment whether this is the case.

https://github.com/magento/magento2/commit/863ec2ebe29bc1c707383f05c37ebcfeb8437f94

What does MC-18193 refer to?

**Comment by ihor-sviziev** [comment]
@ryanpalmerweb, thank you for the confirmation! I'm closing this issue as it was already fixed in https://github.com/magento/magento2/commit/863ec2ebe29bc1c707383f05c37ebcfeb8437f94.
